### PR TITLE
[Do not review] Top of tracing per-param FSDP stack - to be broken up

### PR DESCRIPTION
**Pull request by yf225**
Taking over the FSDP tracing work from https://github.com/pytorch/pytorch/pull/115410. Not all changes are copied from there - only the ones needed for tracing per-param FSDP.

**Dynamo / AOTAutograd code changes:**
1. queue_callback support (new)
2. Guard fix in generic_jump (new)
3. Relax constraint in _mirror_autograd_meta_to to allow requires_grad=True leaf tensor (already PR’ed: https://github.com/pytorch/pytorch/pull/120349) (new)
4. submodule registration on FSDP managed module (maybe hacky, need more thoughts) (copied from Voz's PR)
5. [TODO] Allow `resize_` and no-grad `_foreach_copy_` without issuing clone at beginning of graph.
6. In lieu of functionalization changes above, use custom FX passes to "re-inplace" the inplace ops in the FWD graph and CA graph, to 1) de-duplicate between saved tensors and hook closed-over tensors, 2) reduce `as_strided_scatter / copy_` etc ops and simplify the graph

**ppFSDP code changes:**
1. Move all pre-forward / post-forward / pre-backward / post-backward hooks to global scope in ppFSDP code (known requirement by compiled autograd hook support).
2. Under compile, keep unsharded_param as leaf tensor, and audit all uses of .all_gather_output and replace them with ._unsharded_param (particularly copy_out), so that `._unsharded_param` is graph input and `.all_gather_output` is *not* graph input. This is to avoid "pass multiple views (some of them undifferentiated) of the same tensor to the graph, and mutate some of them" problem which doesn't play well with AOTAutograd.
3. [Can improve] Gate all cuda stream and event usage with is_torchdynamo_compiling() in ppFSDP code. (known necessary change)
5. [Can improve] Under compile, use DTensor.from_local instead of _from_local_no_grad because DTensor() constructor is not supported by Dynamo yet. (new)
7. [Can improve] Under compile, use register_hook to call pre_backward and mimic multi_grad_hook "any" mode behavior. (new)
8. [Can improve] Under compile, ignore RegisterPostBackwardFunction, instead rely on _root_post_backward_final_callback to call the param_group.post_backward(), and rely on Inductor comm reordering to optimally decide when to issue prefetch and reshard.
9. [Hack] t.grad_fn is not supported by Dynamo yet, so we set grad_fns = [] and unconditionally do unshard in _prefetch_unshard() under compile. (new)
10. [Hack] Skip `self._fsdp_param_group.post_forward_mesh_info = None` under compile, because it seems to prevent `.resize_(0)` from issuing at end of forward graph under compile. (new)

**Change to how we run ppFSDP:**
1. [Can improve] Because Dynamo is not able to trace `_lazy_init` of ppFSDP yet, we have to run ppFSDP eager mode once to do the lazy initialization, and then trace ppFSDP. (new)

**If eager ppFSDP unit test fails:**
1. Look for `[EAGER TEST FAIL]` comment in the code.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_state.py [hunk 1]
```diff
@@ -106,10 +188,13 @@ def _lazy_init(self) -> None:
             self._state_ctx.all_states.append(state)
             if state._fsdp_param_group:
                 state._fsdp_param_group.lazy_init()
-        if self._fsdp_param_group:
-            # For the root, do not reshard after forward since for training,
-            # the parameters would be freed and all-gathered immediately
-            self._fsdp_param_group.post_forward_mesh_info = None
+        # NOTE(yf225): I believe this causes issue during compile
+        # (post_forward hook thinks that `post_forward_mesh_info` is None thus not doing resharding and not doing .resize_(0). Is it expected eager behavior?)
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_state.py:192]: This is expected behavior. We choose to not call `resize_(0)` on the root FSDP module's parameters after forward since they would be immediately `resize_()`-d back and all-gathered.

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 2]
```diff
@@ -246,8 +246,9 @@ def init_all_gather_output(
         if self.all_gather_output.numel() > 0:
             return  # already initialized
         all_gather_output_size = torch.Size([all_gather_input_numel * world_size])
+        # NOTE(yf225): here we explicitly make fsdp_param.all_gather_output the leaf tensor (instead of unsharded_param being leaf tensor)
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:249]: I want to be careful about this for the long term solution.

The reason we have a distinction between `all_gather_output` and `unsharded_param` is because for certain extensions, they could be different. For example, if we were to support sharding QLoRA parameters, then there could be multiple `all_gather_output`s, but they will get transformed into a single subclass that is the `unsharded_param`.

However, for the default `torch.Tensor` case, `unsharded_param` should share the same data as `all_gather_output` but have any padding removed.

Comment (yf225) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:249]: Discussed with Andrew offline: making `fsdp_param.all_gather_output` (i.e. padded unsharded) the requires_grad=True leaf tensor and making `unsharded_param` (i.e. unpadded unsharded) a differentiable view into it is okay to do.

**Comment by yf225** [comment]
Superseded by https://github.com/pytorch/pytorch/pull/127239.
